### DISCUSSION TITLE
Tramway Form. Warnings on unknown parameters

### DISCUSSION THREAD

**Discussion by kalashnikovisme** [Purple-Magic/tramway#54] [Ideas]
Example

```ruby
class UserForm < Tramway::BaseForm
  properties :first_name, :last_name
end

params = { email: "something@email.com" }
user_form = tramway_form User.new
```

**NOW**

```ruby
user_form.submit params # returns true
```

**SUGGESTION**

```ruby
user_form.submit params # returns true with the message in a default logger `WARNING: email attribute is not know by UserForm` or something like that
```

**Comment by kalashnikovisme** [comment]
@haukot @tkarpesh @moshinaan WDYT?

**Reply by haukot** [reply]
I like this direction to catch errors more early. But personally I'd like more if it would raise an exception in tests. 
Maybe make an option that will receive a Proc, and call it in such cases?
So it could be configured as users like

**Reply by kalashnikovisme** [reply]
I'm not sure that I understand what you mean.Could you please provide an example

**Reply by haukot** [reply]
Something like this, in an initializer

```ruby
Tramway::BaseForm.warnings_logger = -> (msg) { raise(msg) if Rails.env.test? }
```

**Reply by kalashnikovisme** [reply]
Yeah! Nice idea, I like it. Maybe we should prepare separate methods for tests.

Something like this

*spec/tramway_helper.rb*

```ruby
require 'tramway/test'

Tramway::Test.configure do |config|
  config.form.raise_on_unknown_parameters = true # by default false
  config.form.uknown_parameters_logger = :warning # could be error or info
end
```

**Reply by haukot** [reply]
I like the possibility of sending lambda as it provides maximum flexibility. And there will be no need for methods for different cases. But it's up to you

**Reply by kalashnikovisme** [reply]
I don't like methods very much, could you please provide me with an example as you see this :slightly_smiling_face:

**Reply by haukot** [reply]
https://github.com/orgs/Purple-Magic/discussions/54#discussioncomment-8737562 In a comment above :)

**Reply by kalashnikovisme** [reply]
Do you mean something like that?

```ruby
require 'tramway/test'

Tramway::Test.configure do |config|
  config.form.unknown_parameters_reaction = -> { raise :fuck }
end

```

**Reply by haukot** [reply]
Yes
